code/controllers/subsystems/typing.dm

```diff
@@ -0,0 +1,208 @@
+		if(!islist(entry) || !entry[INDEX_PREFERENCE])
+			continue
+		if(!entry[INDEX_INFLIGHT])
+			update_indicator_state_from_winget(client, entry)
+		else if(world.time < entry[INDEX_TIMEOUT])
+			entry[INDEX_INFLIGHT] = FALSE
+		if(no_mc_tick)
+			CHECK_TICK
+		else if(MC_TICK_CHECK)
+			queue.Cut(1, cut_until)
+			return
+	queue.Cut()
+
+/// Return, generating if necessary, a ckey-indexed list holding typing status.
+/datum/controller/subsystem/typing/proc/get_client_record(client/client)
+	PRIVATE_PROC(TRUE)
+	var/ckey
+	if(istext(client))
+		ckey = client
+	else if(istype(client))
+		ckey = client.ckey
+	else
+		return
+	var/list/entry = status[ckey]
+	if(!entry)
+		entry = new (MAX_INDEX)
+		entry[INDEX_PREFERENCE] = client.get_preference_value(/datum/client_preference/show_typing_indicator) == PREF_SHOW
+		entry[INDEX_INFLIGHT] = FALSE
+		entry[INDEX_TIMEOUT] = world.time
+		entry[INDEX_INPUT_STATE] = FALSE
+		entry[INDEX_VERB_STATE] = FALSE
+		status[ckey] = entry
+	return entry
+
+/// Updates client's preference bool for whether typing indicators should be shown.
+/datum/controller/subsystem/typing/proc/update_preference(client/client, preference)
+	var/list/entry = get_client_record(client)
+	if(islist(entry))
+		entry[INDEX_PREFERENCE] = preference
+		update_indicator(client, entry)
+
+/// Updates client|ckey's verb typing state to new_state.
+/datum/controller/subsystem/typing/proc/set_indicator_state(client/client, state)
+	var/list/entry = get_client_record(client)
+	if(islist(entry))
+		entry[INDEX_VERB_STATE] = state
+		update_indicator(client, entry)
+
+/// Request client's input bar state using winget and updating entry accordingly.
+/datum/controller/subsystem/typing/proc/update_indicator_state_from_winget(client/client, list/entry)
+	PRIVATE_PROC(TRUE)
+	set waitfor = FALSE
+	var/timeout = world.time + INFLIGHT_TIMEOUT
+	entry[INDEX_INFLIGHT] = TRUE
+	entry[INDEX_TIMEOUT] = timeout
+	var/content = winget(client, INPUT_HANDLE, "text")
+	if(timeout != entry[INDEX_TIMEOUT]) // We're stale. Touch nothing.
+		return
+	entry[INDEX_INFLIGHT] = FALSE
+	if(QDELETED(client) || !isliving(client.mob))
+		return
+	entry[INDEX_INPUT_STATE] = match_verbs.Find(content) != 0
+	update_indicator(client, entry)
+
+/// Attempt to update the mob's typing state and indicator according to new state.
+/datum/controller/subsystem/typing/proc/update_indicator(client/client, list/entry)
+	PRIVATE_PROC(TRUE)
+	var/mob/target = client.mob
+	var/display = target.stat == CONSCIOUS && entry[INDEX_PREFERENCE] && (entry[INDEX_INPUT_STATE] || entry[INDEX_VERB_STATE]) && !target.is_cloaked() && isturf(target.loc)
+	if(display == target.is_typing)
+		return
+	if(display)
+		if(!target.typing_indicator)
+			target.typing_indicator = new(target)
+		target.is_typing = TRUE
+		target.typing_indicator.show_typing_indicator()
+	else if(target.typing_indicator)
+		target.is_typing = FALSE
+		target.typing_indicator.hide_typing_indicator()
```

Comment:
You may need some way to deconflict multiple calls to these async procs. In theory you are calling one of the two roughly every half-second, and hotkey say can call them too, so multiple calls might be active at once depending on server load, network timing, etc. It's not clear they will be executed sequentially in the right order, or whether that's even desirable.